spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/Spark3ZOrderUDF.java

```diff
@@ -0,0 +1,239 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.actions;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import java.nio.charset.CharsetEncoder;
+import java.nio.charset.StandardCharsets;
+import org.apache.iceberg.util.ZOrderByteUtils;
+import org.apache.spark.sql.Column;
+import org.apache.spark.sql.expressions.UserDefinedFunction;
+import org.apache.spark.sql.functions;
+import org.apache.spark.sql.types.BinaryType;
+import org.apache.spark.sql.types.BooleanType;
+import org.apache.spark.sql.types.ByteType;
+import org.apache.spark.sql.types.DataType;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.types.DateType;
+import org.apache.spark.sql.types.DoubleType;
+import org.apache.spark.sql.types.FloatType;
+import org.apache.spark.sql.types.IntegerType;
+import org.apache.spark.sql.types.LongType;
+import org.apache.spark.sql.types.ShortType;
+import org.apache.spark.sql.types.StringType;
+import org.apache.spark.sql.types.TimestampType;
+import scala.collection.Seq;
+
+class Spark3ZOrderUDF implements Serializable {
+  private static final int STRING_KEY_LENGTH = 16;
+
+  private static final byte[] PRIMITIVE_EMPTY = new byte[ZOrderByteUtils.BUFFER_SIZE];
+
+  private transient ThreadLocal<ByteBuffer> outputBuffer;
+  private transient ThreadLocal<byte[][]> inputHolder;
+  private transient ThreadLocal<ByteBuffer>[] inputBuffers;
+  private transient ThreadLocal<CharsetEncoder> encoder;
+
+  private final int numCols;
+
+  private int inputCol = 0;
+  private int totalBytes = 0;
+
+  Spark3ZOrderUDF(int numCols) {
+    this.numCols = numCols;
+  }
+
+  private void readObject(java.io.ObjectInputStream in)
+      throws IOException, ClassNotFoundException {
+    in.defaultReadObject();
+    inputBuffers = new ThreadLocal[numCols];
+    inputHolder = ThreadLocal.withInitial(() -> new byte[numCols][]);
+    encoder = ThreadLocal.withInitial(() -> StandardCharsets.UTF_8.newEncoder());
+  }
+
+  private ByteBuffer outputBuffer(int size) {
+    if (outputBuffer == null) {
```

Comment:
Still not clear why not initialize it in readObject?  It just makes an un-initialized ThreadLocal reference right? (it makes that a big difference?)

Reply:
And also, I assume we are expecting concurrent calls , wouldn't we need mt-protection on the check whether buffer is null or not, if we go with this approach (hoping we don't have to)?

Reply:
Yeah no reason we couldn't set it in readObject, I think I just initially put it here before I had implemented the other buffers. 

As for thread execution, the way we generate the UDF means that you get a single "outputBuffer" for every thread. The way Spark should be running this is with a single thread per execution context (task) which means that any individual invocation should actually be in serial. That's why we don't have any protections on any of these buffer references.